contracts/ForwarderFactoryV4.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.20;
+import './ForwarderV4.sol';
+import './CloneFactory.sol';
+
+/**
+ * @title ForwarderFactoryV4
+ * @notice This contract will deploy new forwarder contracts using the create2 opcode
+ */
+contract ForwarderFactoryV4 is CloneFactory {
+  address public implementationAddress;
+
+  /**
+   * @notice Event triggered when a new forwarder is deployed
+   * @param newForwarderAddress Address of the newly deployed forwarder
+   * @param parentAddress Address to which the funds should be forwarded
+   * @param feeAddress Address which is allowed to call methods on forwarder contract alongwith the parentAddress
+   * @param shouldAutoFlushERC721 Whether to automatically flush ERC721 tokens or not
+   * @param shouldAutoFlushERC1155 Whether to automatically flush ERC1155 tokens or not
+   */
+  event ForwarderCreated(
+    address newForwarderAddress,
+    address parentAddress,
+    address feeAddress,
+    bool shouldAutoFlushERC721,
+    bool shouldAutoFlushERC1155
+  );
+
+  /**
+   * @notice Initializes the factory with the address of the current forwarder implementation
+   * @param _implementationAddress Address of the current forwarder implementation
+   */
+  constructor(address _implementationAddress) {
+    implementationAddress = _implementationAddress;
+  }
+
+  /**
+   * @notice Creates a new forwarder contract
+   * @param parent Address to which the funds should be forwarded
+   * @param feeAddress Address which is allowed to call methods on forwarder contract alongwith the parentAddress
+   * @param salt Salt to be used while deploying the contract
+   */
+  function createForwarder(
+    address parent,
+    address feeAddress,
+    bytes32 salt
+  ) external {
+    this.createForwarder(parent, feeAddress, salt, true, true);
+  }
+
+  /**
+   * @notice Creates a new forwarder contract
+   * @param parent Address to which the funds should be forwarded
+   * @param feeAddress Address which is allowed to call methods on forwarder contract alongwith the parentAddress
+   * @param salt Salt to be used while deploying the contract
+   * @param shouldAutoFlushERC721 Whether to automatically flush ERC721 tokens or not
+   * @param shouldAutoFlushERC1155 Whether to automatically flush ERC1155 tokens or not
+   */
+  function createForwarder(
+    address parent,
+    address feeAddress,
+    bytes32 salt,
+    bool shouldAutoFlushERC721,
+    bool shouldAutoFlushERC1155
+  ) external {
+    /// include the parent in the salt so any contract deployed directly relies on the parent address
+    bytes32 finalSalt = keccak256(abi.encodePacked(parent, salt));
```

Comment:
Should the fee address also be part of the final salt?

Reply:
Yes, I think we can add feeAddress as well

Reply:
I think we won't be able to add feeAddress as part of the final salt because if we create forwarders for a base wallet by considering fee address as part of the salt, then the address of the forwarder will be different as compared to when we just consider base address as part of salt (i.e. current implementation). Because of this we won't be able to derive the same forwarder addresses for the same wallet as we have on other chains like Eth, Polygon which is useful for cross chain recoveries.

Reply:
I am not sure I understand the argument against it.

The argument for including it is:
If we don't include the fee address as part of the hash. Then anybody can deploy to this address using a different fee address (and same parent). That could be a security issue if we think that the flushing should only happen via fee address

Reply:
@barath’s point is valid, if somebody initiates a contract creations using a fee address we don’t control and since we don’t allow fee address to be updated at a later point, this will essentially make the receive address not usable to manage from Bitgo. Think we should do it.

Reply:
updated